models/base.py

```diff
@@ -8,47 +8,40 @@
 from sqlalchemy.orm import declarative_base, sessionmaker
 from sqlalchemy_utils import database_exists, create_database
 
+import util.config as config
+
 utcnow = func.timezone("UTC", func.current_timestamp())
 
 
 # this is the root SeeChange folder
 CODE_ROOT = os.path.abspath(os.path.join(__file__, os.pardir, os.pardir))
 
-# this is where the data lives
-# (could be changed for, e.g., new external drive)
-DATA_ROOT = os.getenv("SEECHANGE_DATA")
-if DATA_ROOT is None:  # TODO: should also check if folder exists?
-    DATA_ROOT = os.path.join(CODE_ROOT, "output")
-
-DATA_TEMP = os.path.join(CODE_ROOT, "DATA_TEMP")
-
-# to drop the database use: sudo -u postgres psql -c "DROP DATABASE seechange WITH(force)"
-
-# TODO: check with Rob if he has a preferred way of doing this
-# create database here:
-# url = "postgresql://postgres:postgres@postgres:5432/seechange"
-url = "postgresql://postgres:fragile@seechange_postgres:5432/seechange"
-engine = sa.create_engine(url, future=True)
-if not database_exists(engine.url):
-    create_database(engine.url)
-
-# TODO: we need to decide if we want to make expire_on_commit=False the default
-Session = sessionmaker(bind=engine, expire_on_commit=True)
-
-
+_engine = None
+_Session = None
+    
+        
 @contextmanager
 def SmartSession(input_session=None):
     """
-    Retrun a Session() instance that may or may not
+    Return a Session() instance that may or may not
     be inside a context manager.
 
     If the input is already a session, just return that.
     If the input is None, create a session that would
     close at the end of the life of the calling scope.
     """
+    global _Session, _engine
+
     # open a new session and close it when outer scope is done
     if input_session is None:
-        with Session() as session:
+        if _Session is None:
+            cfg = config.Config.get()
+            url = ( f'{cfg.value("db.engine")}://{cfg.value("db.user")}:{cfg.value("db.password")}'
+                    f'@{cfg.value("db.host")}:{cfg.value("db.port")}/{cfg.value("db.database")}' )
+            engine = sa.create_engine( url, future=True, poolclass=sa.pool.NullPool )
+            
+            _Session = sessionmaker(bind=engine, expire_on_commit=True)
```

Comment:
Why not just initialize these two things at the beginning of the file?

Reply:
So that you can set a different config file at runtime.

If you initialize them in the file, then you have to use a hardcoded or environment-variable session location (or config file).

If you do it this way, then you can call Config.init() with a different config file, and then the results of that config will be picked up here when the session is created.

Reply:
yeah that's pretty good.